### PR TITLE
BUGFIX: Partially repair convert uris implementation according to some 8.3 tests

### PR DESCRIPTION
**Pull request by mhsdesign**
- silently ignore unresolved nodes instead of throwing an exeption
- fix tests "URI to existing asset" by implementing step in 9.0
- reimplement https://github.com/neos/neos-development-collection/pull/3542 which was currently omitted due to merges

**Upgrade instructions**

**Review instructions**

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

File: Neos.Neos/Classes/Fusion/ConvertUrisImplementation.php [hunk 1]
```diff
@@ -183,43 +193,52 @@ function (array $matches) use (&$unresolvedUris, $absolute, $nodeAddress) {
...
                 if ($target === null) {
                     return $linkText;
                 }
-                if (preg_match_all('~target="(.*?)~i', $linkText, $targetMatches)) {
+                // todo merge with "rel" attribute if already existent
+                $relValue = $isExternalLink && $setNoOpener ? 'noopener ' : '';
+                $relValue .= $isExternalLink && $setExternal ? 'external' : '';
+                $relValue = ltrim($relValue);
+                if (str_contains($linkText, 'target="')) {
+                    // todo shouldn't we merge the current target value
```
Comment (mhsdesign) [inline Neos.Neos/Classes/Fusion/ConvertUrisImplementation.php:232]: upsi yes that seems to be a thing. And i have the fear that this problem just got worse not better ^^ https://github.com/neos/neos-development-collection/issues/2962

Comment (bwaidelich) [reply Neos.Neos/Classes/Fusion/ConvertUrisImplementation.php:232]: mh.. that's what happens if you turn a simple "URI converter" to an "URI converter and link attribute converter and link attribute adder" swiss army knife..
I would suggest to extract the latter at some point

File: Neos.Neos/Tests/Behavior/Features/Bootstrap/MiscellaneousTrait.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+use Neos\Media\Domain\Model\Image;
+use Neos\Media\Domain\Repository\AssetRepository;
+use Neos\Utility\ObjectAccess;
+
+/**
+ * Step implementations for tests inside Neos.Neos
+ *
+* @internal only for behat tests within the Neos.Neos package
+*/
+trait MiscellaneousTrait
```
Comment (bwaidelich) [inline Neos.Neos/Tests/Behavior/Features/Bootstrap/MiscellaneousTrait.php]: I gave up on my fight against traits. But a "MiscellaneousTrait" really scares me

**Review by bwaidelich — COMMENTED** [review]
Makes sense IMO, thanks for taking care.
But maybe we can split bugfix from the feature!?

Not to make things more complicated, but I can't really judge on the `replaceLinkTargets()` changes right now, the rest I would +1

File: Neos.Neos/Tests/Behavior/Features/Bootstrap/MiscellaneousTrait.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.Neos/Tests/Behavior/Features/Bootstrap/MiscellaneousTrait.php]: yes but we will always have unstructured feature implementations that should NOT reside in the main feature context.
Thats why i introduced that. It should never reach multiple hundert lines and should be refactored when there are more methods who belong together.

**Comment by mhsdesign** [comment]
> But maybe we can split bugfix from the feature!?

imo nope its a regression to remove a feature that worked previously.

**Comment by mhsdesign** [comment]
> Not to make things more complicated, but I can't really judge on the replaceLinkTargets() changes right now

it seems $someone tried to optimize the convert uris stuff from 8.3 while migrating the code to 9.0
it should behave more performant with the cost of being less accurate and not covering all edgecases.

Maybe we should revert those performance optimization and introduce all those additional preg_match and replaces again for more accuracy. People might already have trouble enough with 9.0 silent bugs like these will make things even harder.

or we find a way to accurately performantly archive the same behaviour as in 8.3

File: Neos.Neos/Tests/Behavior/Features/Bootstrap/MiscellaneousTrait.php (hunk 2 above)
Comment (bwaidelich) [reply Neos.Neos/Tests/Behavior/Features/Bootstrap/MiscellaneousTrait.php]: I totally agree to extract the functionality, but let's call it `AssetTrait` to prevent it to become the new t3lib_div :)

**Comment by bwaidelich** [comment]
> imo nope its a regression to remove a feature that worked previously.

OK, you're right. I just fail to give an +1 for the `replaceLinkTargets()`  – OTOH this is not the nicest code to begin with, and almost any change to it will make it better. And if the new version is closer to the 8.3 behavior I'd say: Let's get this in!

Are you OK with changing the trait name though – it's not a big thing, but also easy to change :)

**Comment by kitsunet** [comment]
+1 for renaming the trait and then getting this in.

**Comment by mhsdesign** [comment]
Hmm could you please diff the 8.3 and new 9.0 convert uris implementation?

im really just about to checkout the 8.3 part again and migrate it without any additional performance improvements, as those will cause bugs.


this is the even more expensive but more accurate 8.3 solution.

https://github.com/neos/neos-development-collection/blob/5f80a56210ba903e484e79a99d82b457a226198f/Neos.Neos/Classes/Fusion/ConvertUrisImplementation.php#L188

**Comment by kitsunet** [comment]
We can probably work on this at another point again, given that this is cached I don't see much problem in having this regex there, also I am sure this won't be in the top 5 of performance problems of rendering 🙈 

tl;dr; fine by me to go back to this.

File: Neos.Neos/Tests/Behavior/Features/Bootstrap/MiscellaneousTrait.php (hunk 2 above)
Comment (mhsdesign) [reply Neos.Neos/Tests/Behavior/Features/Bootstrap/MiscellaneousTrait.php]: Adjusted ;)

**Comment by mhsdesign** [comment]
Okay ill merge this now an prepare a pr where i check all neos fusion objects if they should better be remigrated ;)

**Comment by mhsdesign** [comment]
see https://github.com/neos/neos-development-collection/pull/4760
